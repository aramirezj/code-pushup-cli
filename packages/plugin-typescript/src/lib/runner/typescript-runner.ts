import { access, readFile } from 'node:fs/promises';
// eslint-disable-next-line unicorn/import-style
import { dirname, resolve } from 'node:path';
import {
  type Diagnostic,
  createProgram,
  getPreEmitDiagnostics,
  parseConfigFileTextToJson,
  parseJsonConfigFileContent,
  sys,
} from 'typescript';
import type { TypescriptPluginOptions } from '../config.js';

export async function getDiagnostics(
  options: TypescriptPluginOptions,
): Promise<readonly Diagnostic[]> {
  const { tsConfigPath = 'tsconfig.json' } = options;
  const configPath = resolve(process.cwd(), tsConfigPath);
  const basePath = dirname(configPath);

  try {
    await access(configPath);
  } catch {
    throw new Error(`tsconfig not found at: ${configPath}`);
  }

  const configFile = (await readFile(configPath)).toString();
  const { config: strictConfig } = parseConfigFileTextToJson(
    configPath,
    configFile,
  );
  const parsed = parseJsonConfigFileContent(strictConfig, sys, basePath);

  const { options: opt, fileNames } = parsed;
  if (fileNames.length === 0) {
    throw new Error(
      'No files matched by the TypeScript configuration. Check your "include", "exclude" or "files" settings.',
    );
  }
  const program = createProgram(fileNames, opt);
  return getPreEmitDiagnostics(program);
}
