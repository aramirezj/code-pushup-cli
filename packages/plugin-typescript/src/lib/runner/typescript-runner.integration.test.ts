// eslint-disable-next-line unicorn/import-style
import { basename } from 'node:path';
import { describe, expect } from 'vitest';
import { getDiagnostics, getTsConfiguration } from './typescript-runner.js';

describe('getDiagnostics', () => {
  it('should accept valid options', async () => {
    await expect(
      getDiagnostics({
        tsConfigPath:
          'packages/plugin-typescript/mocks/fixtures/basic-setup/tsconfig.json',
      }),
    ).resolves.not.toThrow();
  });

  it('should return diagnostics array', async () => {
    const res = await getDiagnostics({
      tsConfigPath:
        'packages/plugin-typescript/mocks/fixtures/basic-setup/tsconfig.json',
    });
    expect(res).toHaveLength(4);
    expect(res.at(0)?.code).toBe(2307);
  });

  it('should throw if missing tsconfig path', async () => {
    await expect(
      getDiagnostics({
        tsConfigPath: 'missing-tsconfig.json',
      }),
    ).rejects.toThrow('tsconfig not found at: missing-tsconfig.json');
  });

  it('should throw if no files matched by the TypeScript configuration', async () => {
    await expect(
      getDiagnostics({
        tsConfigPath:
          'packages/plugin-typescript/mocks/fixtures/basic-setup/tsconfig-with-exclude.json',
      }),
    ).rejects.toThrow(
      'No files matched by the TypeScript configuration. Check your "include", "exclude" or "files" settings.',
    );
  });
});

describe('getTsConfiguration', () => {
  it('should accept valid TS config file', async () => {
    const config = await getTsConfiguration({
      tsConfigPath:
        './packages/plugin-typescript/mocks/fixtures/basic-setup/tsconfig.json',
    });

    expect({
      ...config,
      // omitting path details for better snapshots
      fileNames: config.fileNames.map(fileName => basename(fileName)),
      options: {
        ...config.options,
        rootDir: basename(config.options?.rootDir ?? ''),
      },
    }).toMatchSnapshot();
  });
});
