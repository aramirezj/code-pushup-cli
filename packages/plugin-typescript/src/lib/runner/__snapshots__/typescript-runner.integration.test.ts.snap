// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTsConfiguration > should accept valid TS config file 1`] = `
{
  "fileNames": [
    "ts-2307-module-not-fount.ts",
    "ts-2322-strict-null-checks.ts",
    "ts-7006-no-implicit-any.ts",
    "ts-7027-strict-property-initialization.ts",
  ],
  "options": {
    "alwaysStrict": true,
    "configFilePath": undefined,
    "module": 1,
    "noImplicitAny": true,
    "rootDir": "src",
    "strict": true,
    "strictBindCallApply": true,
    "strictFunctionTypes": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": true,
    "target": 2,
  },
}
`;
